### DISCUSSION TITLE
Best practise for RibbonBar and implementation

### DISCUSSION THREAD

**Discussion by MrCoffeeAtGitHub** [canton7/Stylet#281] [Q&A]
Hello,

after developing for years with WPF, Stylet come across for me and I felt in love with it! Thank you so much for you effort and that I can use it. When I find a spot where I could help, I'll be happy to contribute to give a little bit back to the community.

One thing I don't find in the wiki or here at discussion sector is the best practise for following situation. I think is situation is common but may be I don't use the correct search terms for it.

I have a shell view where I introduce a ribbon bar. Here all common buttons are visible and I can call sub functions which are implemented as sub view models. This sub view models have specific buttons / commands which I like to offer as context tabs at the ribbon bar. To achieve this, I implemented a event send by ShellViewModel to ShellView. The ShellViewModel tries to read from ActiveItem.View resources by a magic key the RibbonTabs and forward those by event to ShellView code-behind. Here I change the ribbon bar and add the ribbon tabs from the current sub view. 

Is that a good approach? Do you have a better way?

However, when I click on the button at contextual tabs, the actions are expected at ShellViewModel even the DataContext is set correctly at tab level. How do I say Stylet to look for actions at sub view model with guard methods?

Thank you for your tips and great work again!

**Comment by canton7** [comment]
What you've done sounds fairly sensible, although I'd use a property in the VM and a binding in the View rather than an event. I'd also probably use an attached behavior to move logic out of the View's codebehind. 

I might also have a play with trying to avoid passing the view-specific buttons through the ViewModels, and either have the shell view bind directly into the child view, or use an attached behavior. 

I might also be tempted to define a UserControl subclass which has a property containing the View-specific buttons, rather than making them resources. Or perhaps use an attached property instead. 

I'm not at a computer at the moment: I'll have a play when I am. 

> How do I say Stylet to look for actions at sub view model with guard methods?

Set `s:View.ActionTarget="{Binding}"` or similar, see the Actions wiki page for more details.

**Reply by MrCoffeeAtGitHub** [reply]
Thank you for your reply, canton7.

If you'll find the time, I'd be happy to read about your experience with this.

Meanwhile and thanks to open source, I found `s:View.ActionTarget={Binding}` by myself and I'm happy with this solution. It works but I think there should be a better solution and pattern. I'd be happy to find it.
